static/src/javascripts/lib/robust.js

```diff
@@ -0,0 +1,53 @@
+// @flow
+
+/*
+    Swallows (and reports) exceptions. Designed to wrap around modules at the "bootstrap" level.
+    For example "comments throwing an exception should not stop auto refresh"
+ */
+
+import reportError from 'lib/report-error';
+
+const catchErrors = (fn: Function): ?Error => {
+    let error;
+
+    try {
+        fn();
+    } catch (e) {
+        error = e;
+    }
+
+    return error;
+};
+
+const log = (
+    module: string,
+    error: Error,
+    reporter?: Function = reportError
+): void => {
+    if (window.console && window.console.warn) {
+        window.console.warn('Caught error.', error.stack);
+    }
+
+    reporter(error, { module }, false);
+};
+
+const catchErrorsAndLog = (
+    name: string,
+    fn: Function,
+    reporter?: Function
+): void => {
+    const error = catchErrors(fn);
+
+    if (error) {
+        log(name, error, reporter);
+    }
+};
+
+const catchErrorsAndLogAll = (modules: Array<any>): Array<any> =>
+    modules.map(pair => {
+        const [name, fn] = pair;
+        catchErrorsAndLog(name, fn);
+        return undefined;
```

Comment:
I'm still not convinced this will work correctly. In the [existing implementation](https://github.com/guardian/frontend/blob/d99ec442f7bcbe0d1ade712c757cf83f6fc4f8c9/static/src/javascripts-legacy/lib/robust.js#L46-L48), `makeBlocks` maps over the `modules` array and for each module returns a new function that when executed will call `catchErrorsAndLog` with the context bound to `this`, with the first two parameters applied.

This function is executed in the `forEach` loop in [`enhanced/common`](https://github.com/guardian/frontend/blob/5805db95dc687111731d1111cb6c1abdde3559e1/static/src/javascripts-legacy/bootstraps/enhanced/common.js#L373) and [`enhanced/facia`](https://github.com/guardian/frontend/blob/5805db95dc687111731d1111cb6c1abdde3559e1/static/src/javascripts-legacy/bootstraps/enhanced/facia.js#L116)

Instead of returning `undefined` here, I believe we need to return `catchErrorsAndLog.bind(this, name, fn)`

What do you think?